### PR TITLE
feat: Cloudfront auth: support cognito identity providers

### PR DESCRIPTION
**Pull request by jumi-dev**
By default, the package cdk-cloudfront-authorization uses the Cognito Identity Provider for authentication. It was not possible to set another identity provider like SAML or OpenID Connect.

I've introduced a new variable `identityProviders` in the AuthorizationProps to pass an array of identity provider. This variable is used to set the identity providers in the UserPoolClient.

If no value is specified, the value UserPoolClientIdentityProvider.COGNITO is used which is the current behaviour.

In the Authorization object, I changed the visibility of variable userPoolClient from protected to public. This was necessary to specify a dependency between the UserPoolClient and the UserPoolIdentityProvider like in this example.
```
const azureProvider = new cognito.CfnUserPoolIdentityProvider(this, "Azure", {
   // ...
};
const authorization = new SpaAuthorization(this, 'Authorization_SPA', {
   // ...
   identityProviders: [azureProviderCdk],
};
authorization.userPoolClient.node.addDependency(azureProvider);
```
Without access to variable userPoolClient, CDK creates the UserPoolClient when the IdentityProvider isn't created yet.

Please let me know if you have any questions / comments.

Best regards,
Julian

### CODE REVIEW

File: packages/cdk-cloudfront-authorization/src/authorizations.ts [hunk 1]
```diff
@@ -82,6 +84,8 @@ export abstract class Authorization extends Construct {
 
     this.cookieSettings = props.cookieSettings;
 
+    this.identityProviders = props.identityProviders ?? [UserPoolClientIdentityProvider.COGNITO];
```
Comment (hupe1980) [inline packages/cdk-cloudfront-authorization/src/authorizations.ts:87]: Well done! Cognito should be the default

File: packages/cdk-cloudfront-authorization/src/authorizations.ts [hunk 2]
```diff
@@ -39,20 +39,22 @@ export interface AuthorizationProps {
...
   readonly oauthScopes?: OAuthScope[];
   readonly cookieSettings?: Record<string, string>;
+  readonly identityProviders?: UserPoolClientIdentityProvider[];
 }
 
 export abstract class Authorization extends Construct {
   public readonly redirectPaths: RedirectPaths;
   public readonly signOutUrlPath: string;
   public readonly authFlow: AuthFlow;
+  public readonly userPoolClient: IUserPoolClient;
```
Comment (hupe1980) [inline packages/cdk-cloudfront-authorization/src/authorizations.ts:49]:
```typescript
authorization.userPoolClient.node.addDependency(azureProvider);
```
Do we really have to do that?

I can't find any information about this in the [documentation](https://github.com/aws/aws-cdk/tree/37fdc94a9fdcaf37e02e66d6f1bf3c5501aee1a0/packages/%40aws-cdk/aws-cognito#identity-providers)

**Review by hupe1980 — CHANGES_REQUESTED** [review]
Well done. Supporting other providers is really a great feature.

File: packages/cdk-cloudfront-authorization/src/authorizations.ts (hunk 2 above)
Comment (jumi-dev) [reply packages/cdk-cloudfront-authorization/src/authorizations.ts:49]: I just tried it without specifying the dependency and received this error message:
```
21:42:12 | CREATE_FAILED        | AWS::Cognito::UserPoolClient                    | UserPool2/UserPoolClient
The provider Azure does not exist for User Pool eu-central-1_XXXXXXX. (Service: AWSCognitoIdentityProviderService; Status Code: 400; Error Code: InvalidPar
ameterException; Request ID: xxxxxxxxxxx; Proxy: null)
```

There is an example, where they also added the dependency manually: https://github.com/aws-samples/cdk-cognito-idp/blob/master/lib/cognito-idp-stack.ts
```
// Make sure the user pool client is created after the IDP
userPoolClient.node.addDependency(idp);
```

From my point of view, it makes sence that it is necessary. First, the Identity Provider is specified.
```
const idp = new cognito.UserPoolIdentityProviderFacebook(this, 'FacebookIDP', {
   // ...
}
```
Next, a constant is passed to the UserPoolClient. There is no reference to the IdentityProvider created before. Therefore, CDK can't build the dependency on its own.
`supportedIdentityProviders: [cognito.UserPoolClientIdentityProvider.FACEBOOK]`

Comment (hupe1980) [reply packages/cdk-cloudfront-authorization/src/authorizations.ts:49]: Ah okay. I didn't know that sample. Then it's okay, but we should point it out in the README

**Review by hupe1980 — CHANGES_REQUESTED** [review]
We should point out the need for dependency in the README

File: packages/cdk-cloudfront-authorization/src/authorizations.ts (hunk 2 above)
Comment (jumi-dev) [reply packages/cdk-cloudfront-authorization/src/authorizations.ts:49]: I've enhanced the README file and described "addDependency".

**Review by hupe1980 — APPROVED** [review]
LGTM. Thx!

**Pull request merged**